pkg/apis/pipeline/v1beta1/when_validation.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2020 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"knative.dev/pkg/apis"
+)
+
+var validWhenOperators = []string{
+	string(selection.In),
+	string(selection.NotIn),
+}
+
+func (wes WhenExpressions) validate() *apis.FieldError {
+	if err := wes.validateWhenExpressionsFields(); err != nil {
+		return err
+	}
+	if err := wes.validateTaskResultsVariables(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (wes WhenExpressions) validateWhenExpressionsFields() *apis.FieldError {
+	for _, we := range wes {
+		if err := we.validateWhenExpressionFields(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (we *WhenExpression) validateWhenExpressionFields() *apis.FieldError {
+	if equality.Semantic.DeepEqual(we, &WhenExpression{}) || we == nil {
+		return apis.ErrMissingField(apis.CurrentField)
+	}
+	if err := validateWhenExpressionOperator(&we.Operator); err != nil {
```

Comment:
missing validation on input just like we have it on Operator and Values here, this expression is considered valid otherwise:

```
		name: "missing input",
		wes: []WhenExpression{{
			Operator: selection.In,
			Values:   []string{"foo"},
		}},
```

Reply:
in that case, the input will default to an empty string -- added validation that it's not an empty string but wondering if a user may intentionally want to pass in an empty string